feat(adapters): Add Azure OpenAI support

planetf1 opened the pull request:
<!--
Thank you for your pull request. Please review and complete the sections below.
-->

### Which issue(s) does this pull-request address?

<!--
Please include a link to an issue in the tracker.  The issue describes the problem to be solved.  If there is no issue raised for this PR then either raise one with a summary and description of the problem or add a summary and description of the problem here
-->

Closes: #506

### Description

<!-- Provide a description of the change, pay special attention to describing any breaking changes.  The description describes the resolution to the problem described in the linked issue (or to the problem outlined in this PR). -->

Adds support for Azure OpenAI

Tested
 - azure_openai.py example
 - full example test

The environment variables used are those specified by LiteLLM - these differ from typescript
default model: gpt-4o-mini

Still to do
- [X] Investigate slow performance 
  - Current runtime for the example code is 5m07s - from a trace it appears Azure is throttling the API calls to 1 per minute, as there is exactly 1m between iterations. Suspect this is as-designed, and a quota limit
 
- [X] Test with additional models (limited by quota)

### Checklist

<!-- For completed items, change [ ] to [x]. -->

- [x] I have read the [contributor guide](https://github.com/i-am-bee/beeai-framework/blob/main/CONTRIBUTING.md)
- [x] I have [signed off](https://github.com/i-am-bee/beeai-framework/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-dco) on my commit
- [x] Linting passes: `yarn lint` or `yarn lint:fix`
- [x] Formatting is applied: `yarn format` or `yarn format:fix`
- [x] Unit tests pass: `yarn test:unit`
- [x] E2E tests pass: `yarn test:e2e`
- [x] Tests are included <!-- Bug fixes and new features should include tests -->
- [x] Documentation is changed or added
- [x] Commit messages and PR title follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)

planetf1 commented:
Azure appears to be working reliably, but is slow.

Having analysed the responses (with llm help!)

<google-sheets-html-origin><style type="text/css"><!--td {border: 1px solid #cccccc;}br {mso-data-placement:same-cell;}--></style>
Call | Type | Prompt | Request Time | Response Time | Duration (s) | Prompt Tokens | Completion Tokens | Total Tokens
-- | -- | -- | -- | -- | -- | -- | -- | --
1 | Non-Stream | what states are part of New England? | 12:15:08 | 12:15:52 | 44 | 15 | 50 | 65
2 | Non-Stream | what is the capital of Massachusetts? | 12:15:52 | 12:16:53 | 61 | 14 | 8 | 22
3 | Stream | How many islands make up the country of Cape Verde? | 12:16:53 | 12:17:54 / 12:17:54 | 61 | 19 | 24 | 43
4 | Stream | Produce 3 lines each starting with 'Prefix: ' followed by a sentence and a new line. | 12:18:54 | 12:19:55 / 12:19:55 | 61 | 27 | 48 | 75
5 | Non-Stream | How many islands make up the country of Cape Verde? | 12:17:55 | 12:20:55 | 60 | 63 | 18 | 81

In fact a simple query (via curl) is quick:
```
{"choices":[{"content_filter_results":{"hate":{"filtered":false,"severity":"safe"},"self_harm":{"filtered":false,"severity":"safe"},"sexual":{"filtered":false,"severity":"safe"},"violence":{"filtered":false,"severity":"safe"}},"finish_reason":"stop","index":0,"logprobs":null,"message":{"content":"New England is made up of six states: \n\n1. Maine\n2. New Hampshire\n3. Vermont\n4. Massachusetts\n5. Rhode Island\n6. Connecticut\n\nThese states are located in the northeastern region of the United States.","refusal":null,"role":"assistant"}}],"created":1741352084,"id":"chatcmpl-B8RVsrM8npoRUN1dC3YaMhmiBQznU","model":"gpt-4o-mini-2024-07-18","object":"chat.completion","prompt_filter_results":[{"prompt_index":0,"content_filter_results":{"hate":{"filtered":false,"severity":"safe"},"jailbreak":{"filtered":false,"detected":false},"self_harm":{"filtered":false,"severity":"safe"},"sexual":{"filtered":false,"severity":"safe"},"violence":{"filtered":false,"severity":"safe"}}}],"system_fingerprint":"fp_b705f0c291","usage":{"completion_tokens":50,"completion_tokens_details":{"accepted_prediction_tokens":0,"audio_tokens":0,"reasoning_tokens":0,"rejected_prediction_tokens":0},"prompt_tokens":15,"prompt_tokens_details":{"audio_tokens":0,"cached_tokens":0},"total_tokens":65}}
curl -X POST  -H  -H 'Content-Type: application/json' -d   0.02s user 0.01s system 3% cpu 0.728 total
```
But try the same request straight after:
```
{"error":{"code":"429","message": "Requests to the ChatCompletions_Create Operation under Azure OpenAI API version 2024-10-21 have exceeded token rate limit of your current AIServices S0 pricing tier. Please retry after 51 seconds. Please contact Azure support service if you would like to further increase the default rate limit."}}curl -X POST  -H  -H 'Content-Type: application/json' -d   0.02s user 0.01s system 14% cpu 0.213 total
```

This confirms it's a quota limit. 

I propose we continue and merge.

Tomas2D commented on python/docs/backend.md:
```diff
@@ -49,7 +49,7 @@ The following table depicts supported providers. Each provider requires specific
 | `Amazon Bedrock` |  ✅  |         |  `boto3`| AWS_CHAT_MODEL<br>AWS_ACCESS_KEY_ID<br>AWS_SECRET_ACCESS_KEY<br>AWS_REGION_NAME |
 | `Google Vertex`  |  ✅  |         |  | VERTEXAI_CHAT_MODEL<br>VERTEXAI_PROJECT<br>GOOGLE_APPLICATION_CREDENTIALS<br>GOOGLE_APPLICATION_CREDENTIALS_JSON<br>GOOGLE_CREDENTIALS |
-| `Azure OpenAI`   |    |         | Coming soon! | AZURE_OPENAI_CHAT_MODEL<br>AZURE_OPENAI_EMBEDDING_MODEL<br>AZURE_OPENAI_API_KEY<br>AZURE_OPENAI_API_ENDPOINT<br>AZURE_OPENAI_API_RESOURCE<br>AZURE_OPENAI_API_VERSION |
+| `Azure OpenAI`   |  ✅  |         |  | AZURE_CHAT_MODEL<br>AZURE_API_KEY<br>AZURE_API_BASE<br>AZURE_API_VERSION<br>AZURE_AD_TOKEN<br>AZURE_API_TYPE |
```
Add `_OPENAI`, Azure is not just OpenAI.

planetf1 commented:
Will update to use the same variables as typescript, but also support litellm's as a fallback. Opened https://github.com/i-am-bee/beeai-framework/issues/527 to apply to other providers

planetf1 commented:
Updated
 - Now uses 3 of the same variables as typescript for apikey/endpoint/version
 - Other variables in typescript do not seem to be directly supported in liteLLM
 - Updated docs to match
 - code explicitly fallsback to litellm variables (also, settings have higher priority)
 - restructured/simplified validation
 - added unit test case (in addition to the ChatModel tests)
